# Thème Algorithmique :
# Parcours séquentiel d'un tableau

## Recherche d'une occurrence dans un tableau

Écrire une fonction `recherche(liste, element)` qui renvoie `True` si `element` est dans liste, `False` s'il n'y est pas. 
(bien entendu, interdiction d'utiliser le mot-clé `in`...)

In [1]:
def recherche(liste, element) :
    i=0
    while  i<len(liste) and liste[i]!=element:
        i += 1
    if (i<len(liste) and liste[i]==element):
        return True
    else : return False

def recherche2(liste, element):
    i=0
    trouve=False
    while i<len(liste) and not trouve:
        if liste[i]==element:
            trouve=True
            break
        i = i + 1
    return trouve

In [2]:
assert recherche([1,2,3],3)==True
assert recherche([1],1)==True
assert recherche([1,2,3],4)==False
assert recherche([],1)==False

En fonction du cours de mardi dernier, pouvez vous émettre une hypothèse quant à la complexité de la fonction `recherche`?

Si l'élément recherché est au tout début de la liste, la taille de la liste n'a aucune influence et la complexité est constante. Par contre, si l'élément recherché est le dernier, on a un algorithme de complexité linéaire. On parle alors de **complexité dans le pire des cas**.

Jupyter permet l'utilisation de la fonction magique %timeit, qui mesure le temps moyen d'un code sur un très grand nombre d'exécutions.

Etablissez un tableau des temps d'exécution en temps de la fonction de recherche.


In [3]:
from random import *

def fabrique(n):
    '''
    Fabrique une liste de n éléments en insérant un élément au hasard entre 1 et n
    @params : nombre d'élément à fabriquer
    @return : une liste de n éléments non trié
    '''
    liste = []
    for k in range(n):
        liste.append(randint(0,n))
    return liste

On va appeler la fonction avec %timeit pour estimer un temps moyen d'éxecution.
Faites le avec n valant 1, 10, 100 puis 1000
Qu'observez vous ?

In [2]:
%timeit fabrique(10)

9.55 µs ± 605 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


Quelle est la complexité de la fonction `fabrique` ?  
On a un algorithme de complexité linéaire, c'est à dire en O(n).

Même démarche d'observation avec la fonction `recherche`.
Faites des tests 

In [10]:
%timeit recherche([1],1)

315 ns ± 26.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [15]:
%timeit recherche([1,2,3,4,5,6,7,8,9,10],randint(1,10))

1.9 µs ± 51.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


Puis on va introduire une dose d'aléataoire.

In [16]:
%timeit recherche(fabrique(10),randint(1,10))

13.4 µs ± 859 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [17]:
%timeit recherche(fabrique(100),randint(1,100))

117 µs ± 10.4 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


Qu'observez vous ? Que pouvez vous en déduire de la complexité de recherche ?
Est ce cohérent avec votre hypothèse de départ ?

entre 
%timeit recherche([1,2,3,4,5,6,7,8,9,10],randint(1,10))
et 
%timeit recherche(fabrique(10),randint(1,10))
on devrait avoir des résultats corrélés puisque l'on traite la même quantité de données.
Mais on observe des temps de 1.9 µs à 13.4 µs.
On a un rapport de x7 entre les 2.

L'algorithme est de complexité **linéaire (ici n*n)**.  
On dit aussi qu'il est **d'ordre $n^2$**.  


## Recherche d'extremum dans une liste

## Activité débranchée
Vous disposez des cartes numérotées 3, 2, 4, 6, 9, 1, 5 disposées dans n'importe quel ordre.

Vous disposez aussi d'une carte "support" nommée `m`.  

![](data/deck.png)


1. Exprimer oralement l'algorithme permettant de trouver quelle est la carte portant la valeur maximale.
2. Ecrire cet algorithme en langage naturel.

## Création d'une fonction `maximum(liste)`

*Attention : nous allons recréer une fonction qui, bien sûr, existe déjà en Python, sous le nom `max()`.*

Ecrire l'algorithme en Python.  
Il est expressément demandé de construire une fonction nommée `maximum()`, qui prenne en argument une liste, et qui renvoie le plus grand nombre de cette liste.

**Exemple d'utilisation :**
```python
>>> maximum([3,1,6,2])
>>> 6
```


In [5]:
def maximum(liste):
    if len(liste)==0:
        return None
    else : 
        m = liste[0]
        for k in liste :
            if k > m :
                m = k
        return m              
def maximum2(liste):
    if len(liste)==0:
        return None
    else : 
        m = liste[0]
        for k in range(len(liste)) :
            if liste[k] > m :
                m = liste[k]
        return m  

In [6]:
assert maximum([2,7,5])==7
assert maximum([1])==1
assert maximum([])== None

### Complexité de l'algorithme
  
L'algorithme de recherche de `maximum` est de **complexité linéaire**.

# Calcul de la moyenne des termes d'une liste
Créer une fonction `moyenne()` qui prenne en argument une liste et qui renvoie la moyenne des nombres de cette liste.

In [8]:
def moyenne(liste):
    if len(liste) == 0 :
        return None
    else : 
        somme = 0
        for i in range(len(liste)):
            somme = somme+liste[i]
        return somme//len(liste)

In [9]:
assert moyenne([2,4])==3
assert moyenne([2])==2
assert moyenne([])==None

### Complexité de l'algorithme
  
L'algorithme de recherche de `moyenne` est de **complexité linéaire**.